File: Stack/3634-find-mirror-score-of-a-string/find-mirror-score-of-a-string.cpp
class Solution {
public:
    long long calculateScore(string s) {
        long long sum = 0;
        int n = s.size();
        unordered_map<char, stack<int>> stMap;

        for(int i = 0; i < n; i++) {
            char ch = s[i];
            char chMir = 'z' - ch + 'a';

            if(!stMap[chMir].empty()) {
                sum += i - stMap[chMir].top();
                stMap[chMir].pop();
            } else
                stMap[ch].push(i);
        }

        return sum;
    }
};

File: Stack/3634-find-mirror-score-of-a-string/find-mirror-score-of-a-string.py
class Solution:
    def calculateScore(self, s: str) -> int:
        total = 0
        n = len(s)
        stMap = {}

        for i in range(n):
            ch = s[i]
            chMir = chr(ord('z') - ord(ch) + ord('a'))

            if chMir in stMap and stMap[chMir]:
                total += i - stMap[chMir][-1]
                stMap[chMir].pop()
            else:
                if ch not in stMap:
                    stMap[ch] = []
                stMap[ch].append(i)

        return total
